Polygon annotation

BryonLewis opened the pull request:
The beginning of polygon annotation.  I'm holding back on this PR to not create too massive of a change to the code base.  This at least allows a user to create/edit/save polygon annotations.  The only thing not included is the exporting of them to a csv file or importing them from a csv file.

The basic premise is that a track has a bounds (rectangle) and can also have a polygon represented by a GeoJSON polygon object.  A user can edit at anytime the rectangle or the polygon and there is an interface provided to switch between the two.  

The `EditorMenu.vue` handles the toggling of visibility of the polygon or rectangle layers as well as swapping the edit mode and handling the deletion of points in a polygon.  I created a little integrated help system depending on which mode is selected and the item being editing.  Creation of polygon is different than creation of rectangle, and editing is different as well.

**AnnotationLayer.ts** -This has been modified to accept a type representation being either rectangle or polygon.  It also has some styling changes that are dependent on if it is drawing the other type.  So if you are drawing rectangles while having polygon's enabled it will make the rectangles a lighter/thinner line.

**EditAnnotationLayer** - This now has support for polygon as well as rectangle and point.  There are a few additions/modifications to make it easier to add new editing types.

- Types - now supports different types including rectangle, polygon, point
- Modes - differentiates between 'editing' and 'creation' and exposes this information to outside containers,.
- selectedIndex - ability to select individual vertices to delete them when in polygon editing mode.
- Hovering Index and Clicking - These are some utilities to allow selecting of vertices in a polygon.
- Styling Modifications - the styling has been modified slightly for the editHandles when in polygon mode.  Should make it slightly easier to see things and give us a basis for changing additional items, like maybe changing the `move` handle so it is more recognizable (In polygons the center of the polygon isn't always inside of the polygon or clear and easy to distinguish from a vertex).

**Snackbar**
I've updated the snackbar to include a title if provided and properly support TS for use in the editor menu.  There is a toggle to enable a contextual help which will describe what the modes will do when you switch between them.  Just trying to provide help for switching between playback, editing, and creation for the different types of annotations.

Fixes #163

subdavis commented:
# High level feedback

I haven't done a thorough look-through of the code yet.  These comments are more high-level.  I think the overall UI is very well thought out and easy to use.  Might want to take another look at some of the superficial characteristics of the mode switcher, but in general it's great.

I do have some architectural concerns about the direction this codebase is going.  I'm a little worried that if we aren't very careful, complexity could start to snowball and harm the quality of the project overall.  I don't think this is a result of "bad code" -- and I don't think there's anything particularly wrong with any of what you've done here.  Most of the actual blame I believe is mine: this `useWhatever` pattern isn't scaling as well as I had hoped, and it's starting to get a little painful.

## useAnnotationMode

It isn't clear to me what differentiates the responsibility of `useModeManager` and `useAnnotationMode` -- it sort of seems like these should have been in the same file.   The comment above the default export in `useModeManager` seems to support this.

In general, I think a lot more documentation is needed in `useAnnotationMode` to explain what all the state actually controls.  More detailed comments will be left in the reviewer, no action needed yet.

## EditorMenu

Name is kinda vague.  Half a dozen different components in this app could be named EditorMenu.

## Help system.

I totally agree that a fully mature application like this needs a help system. I really appreciate the attention to the user's needs that you've given here.  However, I think the way that the Help system has been integrated in EditorMenu using the snackbar service is somewhat half-baked and not well suited to combine with this PR.  

IMO, more careful consideration is needed about how a help system can integrate and interact with the entire application (not just the features of annotation mode management).  I also really think there should be much better separation of concerns between actual functionality and the help system: separate state management, separate UI components, and a well-thought-out mechanism for allowing the help system to react to changes in application state without being so tightly coupled.  

For me, EditorMenu was a confusing mix of critical app logic and less important help logic.  Logic and state for the help system should be written such that future changes (and bugs) are isolated to the help system, and cannot easily impact critical app logic.

I'd advocate for removing the whole help system from this PR.  The changes to the snackbar are fine to keep.

## Final thoughts

Don't make any changes yet.  Let's discuss and see if we have similar feelings/concerns.

BryonLewis commented:
> I do have some architectural concerns about the direction this codebase is going. I'm a little worried that if we aren't very careful, complexity could start to snowball and harm the quality of the project overall. I don't think this is a result of "bad code" -- and I don't think there's anything particularly wrong with any of what you've done here. 

Agreed. It makes it much harder for this to be a portable library/tool that can be placed inside of other projects.

> ## useAnnotationMode
> It isn't clear to me what differentiates the responsibility of `useModeManager` and `useAnnotationMode` -- it sort of seems like these should have been in the same file. The comment above the default export in `useModeManager` seems to support this.

I originally had it all in `useModeManager`, but felt that it was getting to be a bit overwhelming in size and scope.  Maybe it can still be in `useModeManager` but implemented as a submodule (I don't know exactly what to call this) so it can be self contained and added and removed as necessary?

> ## Help system.
> I totally agree that a fully mature application like this needs a help system. I really appreciate the attention to the user's needs that you've given here. However, I think the way that the Help system has been integrated in EditorMenu using the snackbar service is somewhat half-baked and not well suited to combine with this PR.

I think I mentioned already I wasn't entirely happy with this system and I didn't want to add to the complexity so I kind of just glued this in as something to give some indication and at least push forward a concept of how it might work.  Also felt it was necessary somewhere in the interface to indicate that right clicking closes a polygon as well as double clicking or clicking the beginning point.  I'm open to any and all other options of how to handle this in a way that is easy to understand for the user.

> For me, EditorMenu was a confusing mix of critical app logic and less important help logic. Logic and state for the help system should be written such that future changes (and bugs) are isolated to the help system, and cannot easily impact critical app logic. This will also facilitate good testing
> I'd advocate for removing the whole help system from this PR. The changes to the snackbar are fine to keep.

I'm all for this and developing a more cohesive help system that can be used across the entire app.

subdavis commented:
Working on some simplifications that I'll submit for your consideration sometime hopefully later today.
